### ISSUE TITLE
use_node doesn't work at all

### ISSUE THREAD

**Issue by zaro** [direnv/direnv#319]
Recently upgraded my direnv , and now every .envrc I used that has use_node stopped working with :

    direnv: Unable to find NodeJS version (v8.9.1) in (/Users/me/.nvm/versions/node/)!
    direnv: error exit status 1

After debugging a bit found that the problem is here :

```shell
563   node_prefix=$(
564     # Look for matching node versions in $NODE_VERSIONS path
565     find "$NODE_VERSIONS" -maxdepth 1 -mindepth 1 -type d -name "$node_wanted*" |
566
567     # Strip possible "/" suffix from $NODE_VERSIONS, then use that to
568     # Strip $NODE_VERSIONS/$NODE_VERSION_PREFIX prefix from line.
569     while IFS= read -r line; do echo "${line#${NODE_VERSIONS%!/(MISSING)}/${node_version_prefix}}"; done |
570
571     # Sort by version: split by "." then reverse numeric sort for each piece of the version string
572     sort -t . -k 1,1rn -k 2,2rn -k 3,3rn |
573
574     # The first one is the highest
575     head -1
576   )
```

specifically 

    while IFS= read -r line; do echo "${line#${NODE_VERSIONS%!/(MISSING)}/${node_version_prefix}}"; 

No matter what settings I have, the ${line#...} expansion always fails, and leaves line as absolute path. Here are some experiments in the command line:

```shell
$ export NODE_VERSIONS=/Users/me/.nvm/versions/node/
$ export node_version_prefix=v
$ echo '/Users/me/.nvm/versions/node/v8.9.1' |  while IFS= read -r line; do echo "${line#${NODE_VERSIONS%/(MISSING)}/${node_version_prefix}}"; done
/Users/me/.nvm/versions/node/v8.9.1
$ echo '/Users/me/.nvm/versions/node/v8.9.1' |  while IFS= read -r line; do echo "${line#${NODE_VERSIONS%/(MISSING)}/}"; done
/Users/me/.nvm/versions/node/v8.9.1
$ echo '/Users/me/.nvm/versions/node/v8.9.1' |  while IFS= read -r line; do echo "${line#${NODE_VERSIONS%/(MISSING)}}"; done
v8.9.1
```
I have removed the ! before (MISSING) because it expands in interactive console. But what you can see is that
```shell 
"${line#${NODE_VERSIONS%/(MISSING)}/${node_version_prefix}}"
"${line#${NODE_VERSIONS%/(MISSING)}/}"
```
Do not trigger expansion correctly while 
```shell
"${line#${NODE_VERSIONS%/(MISSING)}}"
```
does. I observe the same behaviour on both macOs/bash-3.2 and ubuntu/bash-4.3 .

**Comment by zaro** [comment]
What I found as a workaround for me is  to copy the current use_node definition from direnv stdlib , and add it to ~/.config/direnv/direnvrc with 

     node_prefix="${NODE_VERSIONS}/${node_version_prefix}${node_prefix}"

line commented.

**Comment by wilmoore** [comment]
Somehow, I did not get the ping @zimbatm. I found this issue just searching around the issue tracker.

@zaro - what's your `$SHELL`?

**Comment by steve-ross** [comment]
I am working on some additional helpers to use direnv with some of our internal projects. We do a lot of node and typically we use NVM to manage versions. Anyway... I finally put it in a repo yesterday. I'm sure linking to this will show how pitiful my bash skills are however, you can find it here if anyone is interested: https://github.com/steve-ross/direnv-helpers

enables you to put stuff like:
`requires_nvm` in your projects `.envrc` 
Which then reads your `.nvmrc` file, installs the required node version via NVM if it is missing, runs `npm install` (if there isn't a node_modules), does a `use node` etc

**Comment by wilmoore** [comment]
I dig the `__prompt_install_nvm` hooks. I wonder if the idea could be further extended to just handle the node binary installation directly without installing an installer?

I don't have enough free time to explore this much further ATM, but ultimately, direnv has enough tooling to negate the need to support X number of installers (i.e. nvm, n, ...).

If kept to basic `sh` scripting, one would need only install `direnv` and create an `.envrc.` to have a working Node.js environment.

**Comment by steve-ross** [comment]
@wilmoore yeah, I totally get that. Our main reason for creating some of these helpers was that we didn't want to have to copy/paste modify configs for the multiple node projects we work on. It also helps us spin up new developers very quickly.

**Comment by speed-of-light** [comment]
this is what worked for me
```
export NODE_VERSIONS=~/.nvm/versions/node
export NODE_VERSION_PREFIX='v'
use node 10.4
```

direnv version
2.16.0

**Comment by MemoryReload** [comment]
@speed-of-light Yeah, you really save my ass buddy! The `use_node()` function in `stdlib.sh` need these environment variables and the default settings are not right. So we must fix this by ourselves
. Thanks again!

**Comment by ustun** [comment]
Wanted to mention volta as an alternative way to pin node versions in projects(actually it is an alternative to nvm): https://docs.volta.sh/guide/

**Comment by segunadebayo** [comment]
Managed to get this working Volta:

```sh
volta install node (grabs the latest version)
export NODE_VERSIONS=~/.volta/tools/image/node/
export NODE_VERSION_PREFIX=''
use node <latest version>
```

**Comment by lee-pai-long** [comment]
This is not working for me still in `direnv` version `2.32.3`.

I set the `NODE_VERSIONS` correctly and the `NODE_VERSION_PREFIX` to `'v'`. I use the `.node-version` file and put the exact version installed in it without the 'v'.
When  `direnv` executes it fails with the error:
```
Unable to find NodeJS version (20.3.0) in (/home/sma/.nvm/versions/node)!
````

But using `ls` in `/home/sma/.nvm/versions/node` shows that there is a `v20.3.0` directory. Looking at the `use_node` method in `stdlib.sh` file, it seems like the value of the version directory path computated is wrong, but it not shown in the error message so I can't debug it.

The workaround I found was to use the `load_prefix` function directly like this:
```
load_prefix $HOME/.nvm/versions/node/v$(cat ./.node-version)
```

The only caveat is that one need to set the exact version number for it to work.

**Comment by lestephane** [comment]
This line looks extremely suspicious:

```
use_node() {
...
  local node_version_prefix=${NODE_VERSION_PREFIX-node-v}
```

Specifying NODE_VERSION_PREFIX will NOT work in bash (which the shebang says the stdlib is written in).

The proper syntax is `${NODE_VERSION_PREFIX:-node-v}` for bash.

**Comment by cdavie-artium** [comment]
I agree with @lestephane, that bash variable syntax looks wrong to me, and `use_node` is not working for me even with the required env vars set.

Why was [commit d5c29ada](https://github.com/direnv/direnv/commit/d5c29ada6b4d4e80d7e16c86bde58e9cd214b628) made? Has it worked for anyone since that commit? cc @f440, @zimbatm

**Comment by f440** [comment]
The `${name:word}` syntax is supported by shells and is legitimate.

bash: https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Shell-Expansions

> When not performing substring expansion, using the form described below (e.g., ':-'), Bash tests for a parameter that is unset or null. Omitting the colon results in a test only for a parameter that is unset. Put another way, if the colon is included, the operator tests for both parameter's existence and that its value is not null; if the colon is omitted, the operator tests only for existence.

zsh: https://zsh.sourceforge.io/Doc/Release/Expansion.html

> If name is set, or in the second form is non-null, then substitute its value; otherwise substitute word. In the second form name may be omitted, in which case word is always substituted.

The reason for changing from `:-` to `:` in commit d5c29ada6b4d4e80d7e16c86bde58e9cd214b628 is that without this change, we cannot omit the node_version_prefix. Some installation managers store version directories without a prefix, so we need node_version_prefix to be empty, hence this modification is necessary.

```
% bash --version | head -n 1
GNU bash, version 5.2.37(1)-release (x86_64-pc-linux-gnu)

# ${name:-word}
% bash -c 'export NODE_VERSION_PREFIX=; echo ${NODE_VERSION_PREFIX:-node-prefix}'
node-prefix

% bash -c 'export NODE_VERSION_PREFIX=""; echo ${NODE_VERSION_PREFIX:-node-prefix}'
node-prefix

# ${name-word}
% bash -c 'export NODE_VERSION_PREFIX=; echo ${NODE_VERSION_PREFIX-node-prefix}'


% bash -c 'export NODE_VERSION_PREFIX=""; echo ${NODE_VERSION_PREFIX-node-prefix}'


```

It appears that those experiencing issues are using nvm. While I don't use nvm myself, according to the Wiki, you'll need to implement one of the following approaches:

- (1) Using 'use node'
  - Prepare a `.nvmrc` file
  - Add the following to your `.envrc`:
    - `NODE_VERSIONS=$HOME/.nvm/versions/node` # Adjust path according to your environment
    - `NODE_VERSION_PREFIX=v`
    - `use node`
- (2) Define 'use_nvm'
  - Please follow the instructions at https://github.com/direnv/direnv/wiki/Node#using-nvm

I've confirmed that both methods work. Here's the code I used for verification:

```
mkdir work && cd work

Access https://gist.github.com/f440/89b0ed9af913aac240833a3de0f8d6a3 and download the Dockerfile in Raw format

docker build -t direnv-test:latest .
docker run -it --rm direnv-test:latest /bin/bash -l

# Change to the directory and verify that node switches correctly
```

If you still encounter issues, please provide the information needed to reproduce the problem, such as installation paths and configuration file contents (.node-version, .nvmrc, .envrc).

P.S.: As I'm currently not using direnv, I might need some time to investigate or might not be able to answer some questions immediately.
